client/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,789 @@
+		updateCh: make(chan struct{}, 1),
+		heap:     NewVaultClientHeap(),
+		logger:   logger,
+	}
+
+	if !c.config.Enabled {
+		return nil, nil
+	}
+
+	// Get the Vault API configuration
+	apiConf, err := config.ApiConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create vault API config: %v", err)
+	}
+
+	// Create the Vault API client
+	client, err := vaultapi.NewClient(apiConf)
+	if err != nil {
+		logger.Printf("[ERR] vault: failed to create Vault client. Not retrying: %v", err)
+		return nil, err
+	}
+
+	// Set the token and store the client
+	client.SetToken(c.config.Token)
+
+	c.client = client
+
+	return c, nil
+}
+
+// NewVaultClientHeap returns a new vault client heap with both the heap and a
+// map which is a secondary index for heap elements, both initialized.
+func NewVaultClientHeap() *vaultClientHeap {
+	return &vaultClientHeap{
+		heapMap: make(map[string]*vaultClientHeapEntry),
+		heap:    make(vaultDataHeapImp, 0),
+	}
+}
+
+// IsTracked returns if a given identifier is already present in the heap and
+// hence is being renewed.
+func (c *vaultClient) IsTracked(id string) bool {
+	if id == "" {
+		return false
+	}
+
+	_, ok := c.heap.heapMap[id]
+	return ok
+}
+
+// Starts the renewal loop of vault client
+func (c *vaultClient) Start() {
+	if !c.config.Enabled || c.running {
+		return
+	}
+
+	c.logger.Printf("[INFO] vaultclient: establishing connection to vault")
+	go c.establishConnection()
+}
+
+// ConnectionEstablished indicates whether VaultClient successfully established
+// connection to vault or not
+func (c *vaultClient) ConnectionEstablished() bool {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.connEstablished
+}
+
+// establishConnection is used to make first contact with Vault. This should be
+// called in a go-routine since the connection is retried til the Vault Client
+// is stopped or the connection is successfully made at which point the renew
+// loop is started.
+func (c *vaultClient) establishConnection() {
+	// Create the retry timer and set initial duration to zero so it fires
+	// immediately
+	retryTimer := time.NewTimer(0)
+
+OUTER:
+	for {
+		select {
+		case <-c.stopCh:
+			return
+		case <-retryTimer.C:
+			// Ensure the API is reachable
+			if _, err := c.client.Sys().InitStatus(); err != nil {
+				c.logger.Printf("[WARN] vaultclient: failed to contact Vault API. Retrying in %v",
+					c.config.ConnectionRetryIntv)
+				retryTimer.Reset(c.config.ConnectionRetryIntv)
+				continue OUTER
+			}
+
+			break OUTER
+		}
+	}
+
+	c.lock.Lock()
+	c.connEstablished = true
+	c.lock.Unlock()
+
+	// Retrieve our token, validate it and parse the lease duration
+	if err := c.parseSelfToken(); err != nil {
+		c.logger.Printf("[ERR] vaultclient: failed to lookup self token and not retrying: %v", err)
+		return
+	}
+
+	// Begin the renewal loop
+	go c.run()
+	c.logger.Printf("[INFO] vaultclient: started")
+
+	// If we are given a non-root token, start renewing it
+	if c.token.Root {
```

Comment:
Root tokens can have TTLs too. Instead, should look at the token's renewable field, which is stored from `parseSelfToken`.